### DISCUSSION TITLE
.NET Compatible gzuncompress

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#54754] [Q&A] [answered]
I am working with a legacy app that has compressed files using PHP's ```gzcompress``` / ```gzuncompress```.

From a bit of disecting, it looks like ```gzcompress``` returns content compressed with a GZipStream but without the GZip header and footer.

Is there a way to read/write these values in .NET?  Does anyone have any ideas?  I would be fine creating a synthetic header/footer for each record and wrapping the value from PHP inside those bytes but I'm not familiar enough with the GZip format to know what to synthesize.

Any help?

**Comment by svick** [comment, marked as answer]
I believe you can use [`ZLibStream`](https://docs.microsoft.com/en-us/dotnet/api/system.io.compression.zlibstream?view=net-6.0&WT.mc_id=DOP-MVP-5002735) for this, which is a new type in .Net 6.

E.g. consider this PHP code:

```php
print(base64_encode(gzcompress("some test data")));
```

That code prints `eJwrzs9NVShJLS5RSEksSQQAKGMFTw==`. You can then read it in .Net 6 preview like this:

```c#
var inputStream = new MemoryStream(Convert.FromBase64String("eJwrzs9NVShJLS5RSEksSQQAKGMFTw=="));
using var zlibStream = new ZLibStream(inputStream, CompressionMode.Decompress);
using var streamReader = new StreamReader(zlibStream);
string output = streamReader.ReadToEnd();
```

(Your real code probably won't use base64, this was just an easy way for me to test this out.)

**Comment by MarkPflug** [comment]
Are you sure that GZipStream is the right format and not DeflateStream? My understanding was that ZLib, which the [php gzcompress documentation](https://www.php.net/manual/en/function.gzcompress.php) calls out uses the deflate algorithm. I had to deal with something similar recently, decoding [open street map pbf data](https://wiki.openstreetmap.org/wiki/PBF_Format), which is a protocol buffer format that uses zlib to compress blobs internally. I finally got the DeflateStream to decompress the data, but the caveat was that DeflateStream expects the stream to *not* contain a zlib header. The zlib header adds a two byte header to the start and a 4 byte checksum to the end of the data. By trimming these off (`new MemoryStream(data, 2, data.Length - 6)`) the DeflateStream was able to decode the data correctly.

**Reply by svick** [reply]
Trimming the extra bytes and using `DeflateStream` should work, but on .Net 6, just using `ZLibStream` is the easier option.
